sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,221 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "hmac_regs.h"  // Generated.
+#include "sw/device/lib/base/memory.h"
+
+#define HMAC_FIFO_MAX (HMAC_MSG_FIFO_SIZE_BYTES / 32)
+#define HMAC_HASH_SIZE_BITS 256u
+#define HMAC_KEY_REG_COUNT (HMAC_HASH_SIZE_BITS / 32)
+#define HMAC_DIGEST_REG_COUNT HMAC_KEY_REG_COUNT
+
+/**
+ * Read the status register from |hmac|.
+ *
+ * @param hmac The HMAC device to read the status register from.
+ * @return The contents of hmac.STATUS.
+ */
+static uint32_t get_status(const dif_hmac_t *hmac) {
+  return mmio_region_read32(hmac->base_addr, HMAC_STATUS_REG_OFFSET);
+}
+
+/**
+ * Returns the number of entries in the FIFO of |hmac|. If the FIFO is empty,
+ * this function will return 0, and if the FIFO is full, this funciton will
+ * return |HMAC_FIFO_MAX|.
+ *
+ * @param hmac The HMAC device to check the FIFO size of.
+ * @return The number of entries in the HMAC FIFO.
+ */
+static uint32_t get_fifo_entry_count(const dif_hmac_t *hmac) {
+  return (get_status(hmac) >> HMAC_STATUS_FIFO_DEPTH_OFFSET) &
+         HMAC_STATUS_FIFO_DEPTH_MASK;
+}
+
+/**
+ * A helper funciton for calculating |HMAC_FIFO_MAX| - |get_fifo_entry_count()|.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return HMAC_FIFO_MAX - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
```

Comment:
I think a more realistic scenario is having Tock multiplex HMAC requests from processes using different keys. I would expect plain SHA-256 to be a separate DIF?

Reply:
> Is it possible that HMAC could be reconfigured during runtime? As it has two modes, "simple" HASH and "secure" hash, a scenario could potentially be: most of the time use it just as a checksum, sometimes use it as both.
> 
> In that case I think a separate `configure` routine might be beneficial. Could be done similar to UART DIF, where there is a static `configure` function, and two public API `dif_..._init` and `dif_..._configure`? What do you think?

Absolutely, but my thinking is that the SHA256 and HMAC use cases are different enough that it warrants initializing a new `dif_hmac_t`. After all, `dif_hmac_init` takes a `dif_hmac_config_t` so in that sense it _is_ the `configure` routine. If there's a case for finer grain configuration then perhaps they should be added individually, like `dif_hmac_mode_set` or `dif_hmac_key_set`.


> I think a more realistic scenario is having Tock multiplex HMAC requests from processes using different keys. I would expect plain SHA-256 to be a separate DIF?

This would break with the pattern of having a DIF per IP block, and considering there's only a single bit difference between the two modes this seems like overkill. Are there other DIFs that support multiplexing?

Reply:
I think all of these requirements can be handled in the driver level.

Reply:
I had a chat with @gkelly and, upon reflection, we're not sure having more than one `dif_hmac_t` for the same MMIO region is particularly safe (it opens up too many concurrency issues).

This almost certainly suggests, as @silvestrst says, that there should be a way, in this DIF, to:
- choose a different HMAC mode, endianness configuration,
- enable disable both the HMAC datapath and the SHA datapath
- load in a specific secret key

These should be different from initialisation, and preferably different from each other.